### PR TITLE
Canon control, Mechanisms over network and NetworkShildTransform on ShipBlocks

### PR DESCRIPTION
**Pull request by Myrilandel**
#New Features 
- CanonController 
The CanonController is now able to handle only one player at the same time.
Now able to be left by a player

- Mechanisms architecture
Now fully working over the network
Refactored with Commands and RPCs

- MechanismHandler
This script is associated to the player prefab.
Detects when the E key is pressed inside a MechanismActivableDetector.
Attributes the player to the nearest Mechanism.
Also handles the behaviour of the player when he leaves the Mechanism

### CODE REVIEW

**Comment by Myrilandel** [comment]
#New Features 
- CanonController 
The CanonController is now able to handle only one player at the same time.
Now able to be left by a player

- Mechanisms architecture
Now fully working over the network
Refactored with Commands and RPCs

- MechanismHandler
This script is associated to the player prefab.
Detects when the E key is pressed inside a MechanismActivableDetector.
Attributes the player to the nearest Mechanism.
Also handles the behaviour of the player when he leaves the Mechanism

**Comment by Myrilandel** [comment]
##New Features 
- CanonController 
The CanonController is now able to handle only one player at the same time.
Now able to be left by a player

- Mechanisms architecture
Now fully working over the network
Refactored with Commands and RPCs

- MechanismHandler
This script is associated to the player prefab.
Detects when the E key is pressed inside a MechanismActivableDetector.
Attributes the player to the nearest Mechanism.
Also handles the behaviour of the player when he leaves the Mechanism

File: Assets/Scripts/Mechanisms/ActivableMechanism.cs [hunk 1]
```diff
@@ -9,63 +9,46 @@
...
-	 * Launched by the base.Start, we instantiate as a child GameObject an ActivableMechanism that has a ActivableMechanismDetector.
-	 * We notify the ActivableMechanismDetector that we are the Mechanism associated to it.
-	 **/
-    protected override void AttributeMechanismObject()
-    {
-        mechanismObject = (GameObject)Resources.Load("Mechanisms/ActivableMechanism");
-        mechanismObject = Instantiate(mechanismObject, transform);
-        mechanismObject.GetComponent<ActivableMechanismDetector>().SetInetrractableParent(this);
-    }
-
```
Comment (Veltouille) [inline Assets/Scripts/Mechanisms/ActivableMechanism.cs:25]: Ok donc on n'attribut plus dynamiquement l'activable mechanism à l'obj.

File: Assets/Scripts/Mechanisms/Mechanism.cs [hunk 2]
```diff
@@ -9,51 +9,40 @@
...
-    protected virtual void Awake()
-    {
-        AttributeMechanismObject();
-    }
-
-    /** AttributeMechanismObject, protected abstract void
-	 * This method should  be override in child scripts 
-	 * We use it to instantiate a childGameObject to the Mechanism wich handles the behaviour to activate it.
-	 **/
-    protected abstract void AttributeMechanismObject();
```
Comment (Veltouille) [inline Assets/Scripts/Mechanisms/Mechanism.cs:34]: La même chose. On ne place plus l'ActivableMechanism dans le gameObj

File: Assets/Scripts/ManagerGame.cs [hunk 3]
```diff
@@ -35,5 +27,13 @@ public override void OnStartServer()
         ship1.GetComponent<ShipController>().crew = 1;
         ship1.name = "ship1";
         NetworkServer.Spawn(ship1);
+
+        var canon = Instantiate(
+                Canon,
+                Canon.transform.position,
+                Canon.transform.rotation);
+        canon.GetComponent<CanonController>().parentNetId = ship0.GetComponent<NetworkIdentity>().netId;
```
Comment (Veltouille) [inline Assets/Scripts/ManagerGame.cs:35]: Ahhhh, tu get ici le parentID pour le mecanisme.

File: Assets/Scripts/CanonController.cs [hunk 4]
```diff
@@ -7,56 +7,75 @@
...
+    public override void RpcOnActivation(NetworkIdentity activatorID)
     {
-        if (other.tag == "Player" && !isUsed)
-        {
-            other.transform.position = shootPlace.position;
-            other.transform.LookAt(transform);
-            playerUsingCanon = other.GetComponent<PlayerController>();
-            isUsed = true;
-        }
+        activatorID.transform.position = shootPlace.position;
```
Comment (Veltouille) [inline Assets/Scripts/CanonController.cs]: A terme il va falloir homogénéiser tout les noms ID parce que l'ai l'impression que ActivatorID et UserID font référence à la même entité.

File: Assets/Scripts/MechanismHandler.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    /// RPC outside mechanism, is an RPC that will reset the values of the player going out of the mechanism.
+    /// </summary>
+    /// <param name="mechanism">The mechanism.</param>
+    [ClientRpc]
+	private void RpcOutsideMechanism()
+	{
+		UsingMechanism = null;
+		mechanismID = null;
+	}
+
```
Comment (Veltouille) [inline Assets/Scripts/MechanismHandler.cs:56]: Méthodes d'attribution Ok

**Review by Veltouille — APPROVED** [review]
Like said Sirius !

MERGE !!!!!

**Pull request merged**